### DISCUSSION TITLE
New tutorial story?

### DISCUSSION THREAD

**Discussion by nolar** [nolar/kopf#604] [Ideas]
Now, the documentation uses an EphemeralVolumeClaim operator as a [step-by-step tutorial](https://kopf.readthedocs.io/en/latest/walkthrough/problem/). That problem is not relevant anymore — it is solved by K8s itself — via its [ephemeral inline volumes](https://kubernetes.io/blog/2020/01/21/csi-ephemeral-inline-volumes/), or will be solved soon.

The documentation needs a new story to tell. Something simple to understand without deep K8s knowledge, realistic (not fully fictional), and at the same time, complex enough to demonstrate all the key features of Kopf:

* Accessing specs, populating statuses, exchanging information between handlers and resources.
* Change-detecting handlers (creation, update, deletion).
* Silently watching handlers (`@on.event`).
* Daemons, timers.
* Creating children resources (especially, the built-in ones).
* Ease of local execution (e.g. in CLI, PyCharm or similar IDEs).
* Maybe something else (but not all of it).

What that story could be? What is a "good" problem statement?

**Comment by thismatters** [comment]
A realistic problem which isn't solved by a built-in resource is performing a green-blue migration of a versioned application.

* Accessing specs, populating statuses, exchanging information between handlers and resources.
  * specs include version (used as tag for image)
  * status shows current version and will be compared to `spec.version`
* Change-detecting handlers (creation, update, deletion).
  * standard CRUD stuff for app lifecycle
* Silently watching handlers (`@on.event`).
  * watch for label changes to move through migration steps
* Daemons, timers.
  * A daemon to check that no resource has been deleted
  * A timer to roll back the migration when it goes too long 
* Creating children resources (especially, the built-in ones).
  * Application would be a `deployment`, needs `service` and `ingress`, there could be other sidecar `deployment`s (e.g. redis cache)
* Ease of local execution (e.g. in CLI, PyCharm or similar IDEs).
  * this relies on having an application image... which might be troublesome, but there is certainly a suitable boilerplate project image out there.
